### PR TITLE
Add J2CL Lit interaction overlay primitives

### PR DESCRIPTION
**Pull request by vega113**
## Summary
- Adds J2CL Lit primitives for interaction overlay parity: mention suggestions, task metadata, reaction row, reaction picker, reaction authors, and overlay layering.
- Registers the primitives in the Lit shell entrypoint while keeping all mutation/data decisions delegated to parent J2CL models/controllers.
- Adds keyboard, focus, ARIA, close-event, malformed-data, and edge-case Web Test Runner coverage plus a changelog fragment.

Refs #970

## Verification
- `cd j2cl/lit && npm test` — pass, 21 test files / 64 tests.
- `cd j2cl/lit && npm run build` — pass.
- `git diff --check` — pass.
- `python3 scripts/assemble-changelog.py` — pass, assembled 241 entries.
- `python3 scripts/validate-changelog.py --fragments-dir wave/config/changelog.d --changelog wave/config/changelog.json` — pass.

## Review
- Self-review completed.
- Claude Opus review loop completed through Round 5; final review marked the slice ship-ready with no blockers.
- Follow-up notes from review are controller-wiring concerns for later slices: downstream controllers should pass accessible reaction labels and provide outer page inert/backdrop behavior where needed.

## Summary by CodeRabbit

* **New Features**
  * Added several accessible overlay UI components: overlay layer, mention suggestions, task details popover, reaction row/picker/authors with keyboard navigation, focus management, and selection events.

* **Bug Fixes / Improvements**
  * Improved focus trapping/restoration, ARIA labeling/live announcements, single-active picker behavior, and due-date validation UX.

* **Tests**
  * Added comprehensive tests covering keyboard flows, selection/close events, live-region updates, focus behavior, and accessibility.

* **Documentation**
  * Added changelog entry summarizing the new Lit overlay primitives and test coverage.

### CODE REVIEW

File: j2cl/lit/src/elements/reaction-picker-popover.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    if (activePicker === this) {
+      activePicker = null;
+    }
+    super.disconnectedCallback();
+  }
+
+  updated(changed) {
+    if (changed.has("open") && this.open) {
+      if (activePicker && activePicker !== this) {
+        activePicker.close("superseded");
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/reaction-picker-popover.js]: Fixed in `ab80e8829`: `updated()` now clears `activePicker` when `open` flips to false, and `reaction-picker-popover.test.js` covers the property-close path so a later picker does not emit a spurious `superseded` close from the previously closed instance.

File: j2cl/lit/src/elements/mention-suggestion-popover.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    const candidates = this.safeCandidates();
+    const activeIndex = this.clampedIndex(this.activeIndex, candidates);
+    const activeId = candidates.length > 0 ? this.optionId(activeIndex) : "";
+    return html`
+      <div class="popover">
+        <div
+          role="listbox"
+          aria-label="Mention suggestions"
+          aria-activedescendant=${ifDefined(activeId || undefined)}
+        >
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/mention-suggestion-popover.js:83]: Fixed in `cb6a1f796`: the mention listbox is now programmatically focusable, receives focus when opened, and the test asserts that the focused element carries `aria-activedescendant`.

File: j2cl/lit/src/elements/task-metadata-popover.js [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    }
+  }
+
+  render() {
+    if (!this.open) {
+      return "";
+    }
+    const titleId = "task-metadata-title";
+    const errorId = "task-due-date-error";
+    return html`
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/task-metadata-popover.js]: Fixed in `cb6a1f796`: task metadata popovers now allocate per-instance heading/error IDs and the tests cover two simultaneous instances plus the `aria-labelledby` association.

File: j2cl/lit/src/elements/reaction-row.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    const count = Number(value || 0);
+    return Number.isFinite(count) ? Math.max(0, count) : 0;
+  }
+
+  safeReactions() {
+    return Array.isArray(this.reactions) ? this.reactions : [];
+  }
+
+  humanizeName(value) {
+    return String(value || "").replaceAll("_", " ");
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/reaction-row.js]: Fixed in `cb6a1f796`: `humanizeName` now uses `split("_").join(" ")`, which is ES2020-compatible with the Lit bundle target.

File: j2cl/lit/src/elements/reaction-picker-popover.js [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+      }
+      activePicker = this;
+      queueMicrotask(() => this.renderRoot.querySelector("button")?.focus());
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/reaction-picker-popover.js]: Fixed in `ab80e8829` and still present after `cb6a1f796`: `updated()` clears `activePicker` when `open` flips false, with regression coverage for controller-driven property close. Resolving this duplicate stale thread.

File: j2cl/lit/src/elements/mention-suggestion-popover.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/mention-suggestion-popover.js:83]: Fixed in `9a23a9cbe`: the mention listbox is focusable with `tabindex="-1"`, receives focus on open, and tests assert focus plus `aria-activedescendant` ownership.

File: j2cl/lit/src/elements/mention-suggestion-popover.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    return (index + count) % count;
+  }
+
+  optionId(index) {
+    return `${this.optionIdPrefix}-${index}`;
+  }
+
+  safeCandidates() {
+    return Array.isArray(this.candidates) ? this.candidates : [];
+  }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/mention-suggestion-popover.js:197]: Fixed in `9a23a9cbe`: `safeCandidates()` now filters malformed entries before rendering, with regression coverage for null/primitive candidate inputs.

File: j2cl/lit/src/elements/reaction-authors-popover.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    }
+    const authors = this.safeAuthors();
+    const label = this.reactionLabel || this.emoji;
+    return html`
+      <section
+        class="popover"
+        role="region"
+        tabindex="-1"
+        aria-label=${`Authors for ${label}`}
+      >
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/reaction-authors-popover.js]: Fixed in `9a23a9cbe`: reaction authors now falls back to `this reaction` when no reaction label or emoji is available, and the missing-metadata label path is tested.

File: j2cl/lit/src/elements/reaction-row.js [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  safeCount(value) {
+    const count = Number(value || 0);
+    return Number.isFinite(count) ? Math.max(0, count) : 0;
+  }
+
+  safeReactions() {
+    return Array.isArray(this.reactions) ? this.reactions : [];
+  }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/reaction-row.js:138]: Fixed in `9a23a9cbe`: `safeReactions()` now filters malformed entries before rendering, with regression coverage so one bad item does not break the row.

File: j2cl/lit/src/elements/task-metadata-popover.js [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+    this.error = "";
+    this.focusTargetId = "";
+    this.addEventListener("keydown", this.onKeyDown);
+  }
+
+  updated(changed) {
+    if (changed.has("open") && this.open) {
+      queueMicrotask(() => this.renderRoot.querySelector("select")?.focus());
+    }
+  }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/task-metadata-popover.js:86]: Fixed in `9a23a9cbe`: stale task validation state is cleared when the dialog closes, and the close/reopen path is covered by a regression test.

File: j2cl/lit/src/elements/task-metadata-popover.js [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+          Assignee
+          <select name="assignee">
+            <option value="">Unassigned</option>
+            ${this.safeParticipants().map(
+              participant => html`
+                <option
+                  value=${participant.address || ""}
+                  ?selected=${(participant.address || "") === this.assigneeAddress}
+                >
+                  ${participant.displayName || participant.address}
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/task-metadata-popover.js:115]: Fixed in `43144ca3f`: `safeParticipants()` now filters malformed entries before rendering, with regression coverage for null/primitive participant inputs.

File: j2cl/lit/src/elements/task-metadata-popover.js [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+    );
+  }
+
+  safeParticipants() {
+    if (!Array.isArray(this.participants)) {
+      return [];
+    }
+    return this.participants.filter(
+      participant => participant && typeof participant === "object" && !Array.isArray(participant)
+    );
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/task-metadata-popover.js]: Already addressed in `43144ca3f`: `safeParticipants()` now requires `typeof participant.address === "string" && participant.address.trim() !== ""`, so objects without a usable address are filtered out before rendering.

File: j2cl/lit/src/elements/task-metadata-popover.js [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+      >
+        <h2 id=${titleId}>Task details</h2>
+        <label>
+          Assignee
+          <select name="assignee">
+            <option value="">Unassigned</option>
+            ${this.safeParticipants().map(
+              participant => html`
+                <option
+                  value=${participant.address || ""}
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/task-metadata-popover.js:112]: Fixed in `71cd0c480`: added `assigneeFallbackOption()` that renders a selected `<option>` for `assigneeAddress` when it is not present in `safeParticipants()`. A regression test confirms that saving without changing the assignee field preserves the original address when the assignee is absent from the participants list.

File: j2cl/lit/src/elements/mention-suggestion-popover.js [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+    }
+    if (event.key === "ArrowDown" || event.key === "ArrowUp") {
+      event.preventDefault();
+      const offset = event.key === "ArrowDown" ? 1 : -1;
+      this.activeIndex = this.clampedIndex(this.activeIndex + offset);
+      return;
+    }
+    if (event.key === "Enter") {
+      event.preventDefault();
+      this.selectCandidate(this.clampedIndex(this.activeIndex));
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/mention-suggestion-popover.js]: Fixed in `f41527111`: the Enter handler now guards with `safeCandidates().length > 0` (matching the existing Tab branch) before calling `preventDefault()`, so Enter passes through to the host editor when the suggestions list is empty.

[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/mention-suggestion-popover.js]: Fixed in `f41527111`: Enter is only consumed when the mention popover has selectable candidates. Follow-up `956f22fa1` keeps the fix regression-pinned with empty/candidate Enter tests after the conflict-resolution pass.

File: j2cl/lit/src/elements/mention-suggestion-popover.js [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+      })
+    );
+  }
+
+  clampedIndex(index, candidates = this.safeCandidates()) {
+    const count = candidates.length;
+    if (count === 0) {
+      return 0;
+    }
+    return (index + count) % count;
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/mention-suggestion-popover.js]: Fixed in `ee8183395`: `clampedIndex()` now uses normalized modulo so indexes below `-candidateCount` wrap to a valid option. Added a regression test for `active-index="-3"` with two candidates; local Lit suite passed with 79 tests, plus build/diff/changelog checks.

File: j2cl/lit/src/elements/task-metadata-popover.js [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+      return [];
+    }
+    return this.participants.filter(
+      participant =>
+        participant &&
+        typeof participant === "object" &&
+        !Array.isArray(participant) &&
+        typeof participant.address === "string" &&
+        participant.address.trim() !== ""
+    );
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/task-metadata-popover.js]: Fixed in `50dd072f8`: `safeParticipants()` now chains a `.map()` after the filter, trimming each `address` before returning so rendered option values and submitted `assigneeAddress` are always canonical (no surrounding whitespace). Regression test covers the padded-address submit path.

File: j2cl/lit/src/elements/interaction-overlay-layer.js [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+
+  focusableControls() {
+    const slot = this.renderRoot.querySelector("slot");
+    const assigned = slot?.assignedElements({ flatten: true }) || [];
+    const controls = [];
+    for (const element of assigned) {
+      if (this.isFocusable(element)) {
+        controls.push(element);
+      }
+      controls.push(...Array.from(element.querySelectorAll(this.focusableSelector())));
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/interaction-overlay-layer.js]: Fixed in `50dd072f8`: `focusableControls()` now delegates to `collectFocusableControls()` which recurses into `element.shadowRoot`, collecting focusable controls from nested shadow trees. Active element resolution uses `composedPath()[0]` first so the current index is correctly located across shadow boundaries. Regression test: a slotted custom element with shadow-DOM buttons wraps Tab focus in both directions.

File: j2cl/lit/src/elements/task-metadata-popover.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/task-metadata-popover.js]: Fixed in `50dd072f8`: `safeParticipants()` now returns cloned participants with `address` trimmed, and the regression test asserts the rendered select value and submitted `assigneeAddress` are canonical. Local verification: targeted affected tests passed, full Lit suite passed with 81 tests, build/diff/changelog checks passed.

File: j2cl/lit/src/elements/interaction-overlay-layer.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/interaction-overlay-layer.js]: Fixed in `50dd072f8`: the modal focus trap now collects focusables inside slotted custom elements’ shadow roots and uses the composed event target as the active control. Added a regression test that wraps focus both directions through shadow-root buttons. Local verification: targeted affected tests passed, full Lit suite passed with 81 tests, build/diff/changelog checks passed.

File: j2cl/lit/src/elements/reaction-picker-popover.js [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+    }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/reaction-picker-popover.js:68]: Fixed in current branch: `updated()` now clears `activePicker` when `open` flips to `false`, so a parent closing the picker declaratively no longer leaves a stale active reference.

File: j2cl/lit/src/elements/task-metadata-popover.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/task-metadata-popover.js]: Fixed in current branch: `safeParticipants()` now filters entries that lack a non-empty string `address`, so objects with missing or blank addresses are stripped before rendering and cannot produce blank assignee options.

File: j2cl/lit/src/elements/mention-suggestion-popover.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/mention-suggestion-popover.js]: Fixed in current branch: the Enter handler now guards with `this.safeCandidates().length > 0` before calling `preventDefault()`, so Enter is only consumed when there are active suggestions and passes through to the host editor when the list is empty.

File: j2cl/lit/src/elements/task-metadata-popover.js [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+      event.preventDefault();
+      first.focus();
+    }
+  }
+
+  submit = (event) => {
+    event?.preventDefault();
+    const dueDate = this.renderRoot.querySelector("input[name='dueDate']")?.value || "";
+    if (dueDate && !this.isValidDueDate(dueDate)) {
+      this.error = "Use YYYY-MM-DD for the due date.";
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/task-metadata-popover.js]: Fixed in `8eda48f0`: due-date submit now trims the input once before validation and emits the same canonical value. Added a regression test for whitespace-padded valid due dates. Local verification: targeted mention/task tests passed, full Lit suite passed with 83 tests, plus build/diff/changelog checks.

File: j2cl/lit/src/elements/mention-suggestion-popover.js [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+      })
+    );
+  }
+
+  clampedIndex(index, candidates = this.safeCandidates()) {
+    const count = candidates.length;
+    if (count === 0) {
+      return 0;
+    }
+    return ((index % count) + count) % count;
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/mention-suggestion-popover.js]: Fixed in `8eda48f0`: `clampedIndex()` now coerces the incoming index through `Number()` and falls back to `0` for non-finite values before modulo wrapping. Added a regression test for `Number.NaN` active indexes. Local verification: targeted mention/task tests passed, full Lit suite passed with 83 tests, plus build/diff/changelog checks.

File: j2cl/lit/src/elements/task-metadata-popover.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/task-metadata-popover.js]: Fixed in `8bdfcdfe5`: `safeParticipants()` now requires `address` to be a non-empty trimmed string, so address-less objects are dropped before rendering options.

File: j2cl/lit/src/elements/mention-suggestion-popover.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/mention-suggestion-popover.js]: Fixed in `f41527111`: Enter now only calls `preventDefault()` and selects a candidate when `safeCandidates().length > 0`, matching the existing Tab branch guard.

File: j2cl/lit/src/elements/interaction-overlay-layer.js [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+    `;
+  }
+
+  onKeyDown = (event) => {
+    if (!this.open) {
+      return;
+    }
+    if (event.key === "Escape") {
+      event.preventDefault();
+      this.close("escape");
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/interaction-overlay-layer.js:77]: Fixed in `05991e24d`: `interaction-overlay-layer` now ignores Escape when a nested/slotted child already called `preventDefault()`, preventing duplicate close events. Added a regression test that verifies the layer remains open and emits no `overlay-close`. Local verification: targeted affected tests passed, full Lit suite passed with 85 tests, plus build/diff/changelog checks.

File: j2cl/lit/src/elements/mention-suggestion-popover.js [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+    return `${this.optionIdPrefix}-${index}`;
+  }
+
+  safeCandidates() {
+    if (!Array.isArray(this.candidates)) {
+      return [];
+    }
+    return this.candidates.filter(
+      candidate => candidate && typeof candidate === "object" && !Array.isArray(candidate)
+    );
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/mention-suggestion-popover.js]: Fixed in `05991e24d`: mention candidates are now filtered to non-empty string addresses and normalized with trimmed addresses before rendering/selecting. Added a regression test for missing, blank, and padded candidate addresses. Local verification: targeted affected tests passed, full Lit suite passed with 85 tests, plus build/diff/changelog checks.

File: j2cl/lit/src/elements/task-metadata-popover.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/task-metadata-popover.js]: Fixed: `safeParticipants()` now requires `typeof participant.address === "string"` and `participant.address.trim() !== ""` before an entry passes filtering, so objects without a usable address are excluded before rendering.

File: j2cl/lit/src/elements/mention-suggestion-popover.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/mention-suggestion-popover.js]: Fixed: the Enter branch is now guarded with `&& this.safeCandidates().length > 0` (same as Tab), so `preventDefault` is only called when there are candidates to select. Pressing Enter on an empty list passes the event through to the host editor.

File: j2cl/lit/src/elements/task-metadata-popover.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/task-metadata-popover.js]: Fixed in `8bdfcdfe5`: `safeParticipants()` now requires a non-empty string `address` on every entry before it passes the filter, so objects without a usable address are excluded and cannot produce blank assignee options.

File: j2cl/lit/src/elements/mention-suggestion-popover.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/mention-suggestion-popover.js]: Fixed in `f41527111`: the Enter handler now guards with `this.safeCandidates().length > 0` before calling `preventDefault()`, so keypresses are not consumed when the suggestions list is empty.
